### ISSUE TITLE
DLL Detach Handling

### ISSUE THREAD

**Issue by Sunderous** [veeenu/hudhook#30]
Hey, first off kudos for this project, I went down the rabbit hole of DX hooking and using this lets me focus on the more fun parts of what I want to create.

I noticed that when I unload the DLL it crashes my game, and looking at your source there is functionality to unhook, but there's no logic implemeted for DLL Reason = Detach.

Would it make sense to do it that way? Have the hudhook macro generate a reason == DLL_DETACH block that just calls unhook for whatever DX version we're using?

It would make iterating on my code a lot faster than restarting the game every time haha.

Thanks!

**Comment by Sunderous** [comment]
Well I'm just a bit impatient I guess haha, after like 10 more minutes looking at the code I found your eject function:

https://github.com/veeenu/hudhook/blob/8afac9c480c6f243f1b4d119daef48cc1f49dda6/src/lib.rs#L175-L186

Putting it there for anyone else who might stumble across this. Might be worth adding that to the readme example so it's front and center, but all good on my end! Thanks again.

**Comment by Sunderous** [comment]
Sorry, going to re-open this as I think there's still a problem.

Right now, if I inject -> eject() it's all good. But if I try to re-inject the same process again it crashes the second time.

So there may still be something that the eject() cleanup isn't doing correctly, that makes installing the hook a second time crash >.<

**Comment by veeenu** [comment]
Hey, thank you for your feedback!

Inject and using the eject method are the correct thing to do. For an example of usage, check the [practice tool](https://github.com/veeenu/eldenring-practice-tool/blob/main/practice-tool/src/lib.rs#L204); you want to invoke the function from the same thread that is doing the rendering, just to be sure.

If the process crashes again on second injection, I suggest you check where the crash is happening with a debugger like WinDbg or x64dbg. I suggest you do that with an empty render loop, i.e. something that just renders some static text, to rule out non-idempotent memory modifications you may have done to your target process (e.g. non-reverted injections in code caves).

If you can determine `hudhook` is itself the cause of crashing for your game, please gather and report here as much information as possible so I can look into it.

Thank you!

**Comment by Sunderous** [comment]
All good, I don't know if I have another DX11 application handy to see if I can replicate it with. But I'll do a little more analysis first, and close this for now since I can't prove if it's the app or hookhud yet. Thanks.

**Comment by Sunderous** [comment]
So instead of wiring up a debugger, I grabbed some simple injection code and wired that up instead of using an off the shelf injector.

And doing it that way, I can't re-produce it, so I have to assume it was something the other injector was doing, all good in your code!

**Comment by Sunderous** [comment]
Gah, sorry but it still feels like there's something odd going on here, I might need some help diving deeper into it.

Right now I can re-inject as much as I want no problem.

But as soon as I change the DLL code, and re-compile, the next injection fails/crashes the host process 100% of the time. Doesn't matter what the change is. I can restart the app, inject the same DLL that just crashed it, and it'll work until I tweak the code again and re-compile. Which makes it impossible to iterate.

I'm fairly confident the crash is happening when it tries to setup the DX11 hook after being recompiled, the simple log trace just stops here:

![image](https://user-images.githubusercontent.com/24513478/193092711-70446720-c424-4888-aa40-69f5e04a7142.png)

And x64 dbg reports the last error as 0x582 = ERROR_CLASS_ALREADY_EXISTS:

![image](https://user-images.githubusercontent.com/24513478/193092865-c0d27ccb-2549-46f5-8199-0f77ff270332.png)

I don't see that error show up at all during successful injections, so it seems unique to the crash.

But I'm still pretty new to reversing, so I'm not sure how to correlate what would change between compiles (other than binary size) that might impact re-installing the DX11 hook.

If you've got any suggestions where to poke at either in x64 debug or the code itself I can try and get some more details.

Thanks.

**Comment by Sunderous** [comment]
So my guess is that for some reason it's this dummy window class:
https://github.com/veeenu/hudhook/blob/0ddab449018a042662cf6849e7b0403511c22304/src/hooks/dx11.rs#L302-L345

It would make sense that was getting thrown because the DUMMY class was never unregistered, but it doesn't explain why it would only happen after re-compiling. I'd expect it would happen on any injection after the first one.

**Comment by Sunderous** [comment]
Yep, seems to have done the trick if you ignore the traces/commented out line in my fork here:
https://github.com/Sunderous/hudhook/blob/b4cad7822fd55359a46c7ba6fd63134ef6ef0eb9/src/hooks/dx11.rs#L302-L400

If I move the instance handle up to the function's scope so it's available for cleanup later, and move the dummy class name into a single spot just to be safe, then it lets me call UnregisterClassA(class_name, hinstance) after destroying the window.

This appears to have solved my  issue, I can tweak the code, re-compile, and inject as much as I want.

**Comment by veeenu** [comment]
Oh, great catch! So, if I understood correctly, the only relevant difference is unregistering the class? This could actually be a fix for other renderers as well, that probably also aren't unregistering the dummy window class.

**Comment by Sunderous** [comment]
Correct. That's just how I got it to work, but as long as you have a way of getting the hInstance from the window, so you can destroy window -> unregister class with:

DestroyWindow(hWnd)
UnregisterClassA(className, hInstance)

That's all that matters as far as I can tell.

I read somewhere the ordering matters, that you have to destroy first before unregistering the class. But I wasn't using the right build when I had the order reversed, or when I had some code that tries to pull the hInstance out of the window handle, so I can't say for sure it matters haha.

**Comment by veeenu** [comment]
Great! I've renamed the issue and will keep it open to track the necessary modification to all DX hooks. Thank you!

**Comment by veeenu** [comment]
Change applied to DX11 and DX12 hooks.

**Comment by etra0** [comment]
Oh I'm gonna test this as well!

I noticed the same behavior, I can deinject and inject as many times I want, but as soon as I recompile it's a guaranteed crash. 

I'll see if this fixed this issue!

**Comment by veeenu** [comment]
I've had another idea which I'm not 100% sure is going to pan out -- I'm going to try and create a swap chain straight on the desktop HWND. This could be either a disaster or a huge success, but would get rid of a lot of code which I would very much like.

**Comment by veeenu** [comment]
@etra0 @Sunderous turns out my idea might work after all!

Would you guys mind giving [this branch](https://github.com/veeenu/hudhook/tree/desktop-swapchain) a spin?

It works for me both with Dark Souls III and Elden Ring. I'm happy because I can trim away so much code 😅

**Comment by etra0** [comment]
I'll give it a spin this afternoon!

**Comment by etra0** [comment]
I did a quick test with for my tool for The Witcher 3 ([litcher](https://github.com/etra0/litcher)) using a [branch](https://github.com/etra0/hudhook/tree/test-branch) with my changes rebased on your branch and everything seems to keep working. I tested making changes three times, and I also tested deinjecting a debug and then injecting a release version, and it works. Previously that would simply crash at the second reinjection

**Comment by Sunderous** [comment]
Hey, sorry I had to switch to the DX9 hook explicitly for my use case, and then tore that down to just the present hook because I was having a lot of issues trying to stabilize the repeated injection behavior (may not have been necessary, but when in doubt simplify haha).

So I probably don't have a good test case for this right now, but I'll look at the code and see if I can implement the desktop approach in my fork for dx9 and let you know how it goes.

**Comment by veeenu** [comment]
> So I probably don't have a good test case for this right now, but I'll look at the code and see if I can implement the desktop approach in my fork for dx9 and let you know how it goes.

I applied the change to Dx9 as well, so it's all the more welcome if you can test that against Dx9 since I don't have a use case for it myself 😅

**Comment by Sunderous** [comment]
So I threw the new code in there for finding the function addresses with GetDesktopWindow, switched to your detour repo to make that compile again (I'm guessing detour 2 added a breaking change, or just wouldn't work with the new approach?).

And seems to inject/re-inject fine so dx9 working as far as I can tell haha.

**Comment by veeenu** [comment]
Thank you! 😊

New Rust nightly has a change that breaks detours-rs, I opened a PR [here](https://github.com/lostinspiration/detour-rs/pull/1) to fix that.
